### ISSUE TITLE
Consider using bindgen

### ISSUE THREAD

**Issue by FranklinChen** [KWARC/rust-libxml#19]
It might be useful to use [bindgen](https://rust-lang-nursery.github.io/rust-bindgen/) to generate an initial and comprehensive set of raw Rust bindings (in place of the incomplete hand-written bindings) and then refactor the lifecycle management wrappers to go around those.

**Comment by dginev** [comment]
That definitely sounds like an easy win to go up in coverage. Would be a great incentive to add more tests too.

Will see if I get some time to do prep it as an add-on one of these days, but anyone reading this is most welcome to beat me to it :+1:

**Comment by triptec** [comment]
It's funny, I just read up on bindgen remembering something about it from a this-week-in-rust and was about to ask if it was something you had considered.

So yeah it would be nice to have bigger coverage, but how do we also get a good api over those generated bindings? Now I'm not used to C and I do not much care for the documentation at http://xmlsoft.org/html/index.html (though the content is good!) but taking a node from one Document and add it to a node of another Document will result in double free and I guess there's more where that came from. So first I guess the question is should the generated bindings be public? And how do we design the ergonomic api and should that even be another crate? Also I believe this crate has something like 2500 downloads, ie. do we need to keep backwards compatibility?

**Comment by dginev** [comment]
So the good news is there are many well-developed and well-supported libxml wrappers in other languages we can learn from. XML::LibXML in Perl, Nokogiri in ruby, ...list goes on.

And I think the important lessons are:
 - Keep the C-facing interface internal and private,
 - Design a new high level XML-manipulation interface that follows "the spirit" of the high level programming language
 - Expose/document/test the new high level interface for the wrapper consumers.

When I started this wrapper I needed a direct port of Perl's XML::LibXML methods, which has played a role in how the basic API was initially structured. But we never had a high-level API "plan" per se.

It may be a good set of steps to:
 - Run bindgen and expose the resulting C bindings as an internal interface for the crate
 - Write up (and do a few iterations of editing) on a high level API document for the Rust side of things
 - Implement and test the now documented Rust-facing public API

Of course there is no need to get it right at the first try, and I think being at version `0.1.0` there are no guarantees for backwards compatibility -- I would consider the interface largely experimental. My usual instinct would be to stabilize a `1.0` release and guarantee it is backwards compatible until `2.0` etc.

**Comment by FranklinChen** [comment]
I don't have time to contribute fully right now, but I had privately already used bindgen for my own quick and dirty use in place of using rust-libxml. I created a `wrapper.h` file

```
#include <libxml/DOCBparser.h>
#include <libxml/HTMLparser.h>
#include <libxml/HTMLtree.h>
#include <libxml/SAX.h>
#include <libxml/SAX2.h>
#include <libxml/c14n.h>
#include <libxml/catalog.h>
#include <libxml/chvalid.h>
#include <libxml/debugXML.h>
#include <libxml/dict.h>
#include <libxml/encoding.h>
#include <libxml/entities.h>
#include <libxml/globals.h>
#include <libxml/hash.h>
#include <libxml/list.h>
#include <libxml/nanoftp.h>
#include <libxml/nanohttp.h>
#include <libxml/parser.h>
#include <libxml/parserInternals.h>
#include <libxml/pattern.h>
#include <libxml/relaxng.h>
#include <libxml/schemasInternals.h>
#include <libxml/schematron.h>
#include <libxml/threads.h>
#include <libxml/tree.h>
#include <libxml/uri.h>
#include <libxml/valid.h>
#include <libxml/xinclude.h>
#include <libxml/xlink.h>
#include <libxml/xmlIO.h>
#include <libxml/xmlautomata.h>
#include <libxml/xmlerror.h>
#include <libxml/xmlexports.h>
#include <libxml/xmlmemory.h>
#include <libxml/xmlmodule.h>
#include <libxml/xmlreader.h>
#include <libxml/xmlregexp.h>
#include <libxml/xmlsave.h>
#include <libxml/xmlschemas.h>
#include <libxml/xmlschemastypes.h>
#include <libxml/xmlstring.h>
#include <libxml/xmlunicode.h>
#include <libxml/xmlversion.h>
#include <libxml/xmlwriter.h>
#include <libxml/xpath.h>
#include <libxml/xpathInternals.h>
#include <libxml/xpointer.h>
```

and a `build.rs` file

```
extern crate bindgen;

use std::env;
use std::path::PathBuf;

fn main() {
    // Tell cargo to tell rustc to link the system bzip2
    // shared library.
    println!("cargo:rustc-link-lib=xml2");

    // The bindgen::Builder is the main entry point
    // to bindgen, and lets you build up options for
    // the resulting bindings.
    let bindings = bindgen::Builder::default()
        // The input header we would like to generate
        // bindings for.
        .header("wrapper.h")
        // Homebrew location of libxml2 headers.
        .clang_arg("-I/usr/include/libxml2")
        // Finish the builder and generate the bindings.
        .generate()
        // Unwrap the Result and panic on failure.
        .expect("Unable to generate bindings");

    // Write the bindings to the $OUT_DIR/bindings.rs file.
    let out_path = PathBuf::from(env::var("OUT_DIR").unwrap());
    bindings
        .write_to_file(out_path.join("bindings.rs"))
        .expect("Couldn't write bindings!");
}
```

and that worked for me on MacOS for what I needed. I agree that for real use, the generated bindings should be internal and then wrapped with a safe public API.

**Comment by dginev** [comment]
Cool, thanks for sharing your work! Will speed us up to cruising velocity

**Comment by triptec** [comment]
Yes, I would like to comment out the bulk of the imports though and start with like the areas that has already been somewhat covered. Having spend the last 4 years mainly with ruby I'm a big fan of Nokogiri, for me that would be a great choice of inspiration and with some extra functionality that rusts ammonia crate provides.

Though I've mainly been in the business of taking xml and changing it and that's my current use case it would be nice in the future having something like nokogiri's dsl for creating xml/html.

About the plan I do agree.

**Comment by triptec** [comment]
Though it's quite ambitious, the plan I mean but it doesn't include implement all stuff libxml have done..

**Comment by dginev** [comment]
Well, `rust-libxml` needs to be a respectable crate, it is deserving of an ambitious plan :+1: We can keep adding high-level API coverage as we go, one minor version at a time, so no need for the process to feel overwhelming. Nokogiri has a fun feel to it, just writing some now myself (Ruby on Rails at work right now).

**Comment by triptec** [comment]
Looking at the Nokogiri docs, what do you think would be the most idiomatic what to translate a signature like this?
.parse(thing, url = nil, encoding = nil, options = ParseOptions::DEFAULT_XML, &block) ⇒ Object
I'm mainly thinking about the default arguments. As for "thing" I guess something like this (https://medium.com/@jreem/advanced-rust-using-traits-for-argument-overloading-c6a6c8ba2e17) would fit no?

**Comment by triptec** [comment]
Builder pattern?

**Comment by dginev** [comment]
There are two patterns I have been happy in Rust w.r.t default arguments:
 1. Implementing the Default trait and using `..Trait::default()` in initalization
 2. Using macros for flexible signature arguments

In the first approach, you would group all options (that logically group together) into a dedicated say `ParseOptions` struct - for this example - and then pass in any arguments you need via e.g.
```rust
 ParseOptions{ url: "example.com", ..ParseOptions::default()}
```

Then you can abbreviate that via a dedicated macro, if too annoying.

In the second approach you would define a waterfall of macros somewhat on the lines of:
```rust
macro_rules! parse {
  ($thing:expr) => (parse($thing, None, None, ParseOptions::DEFAULT_XML)),
  ($thing:expr, $url:expr) => (parse($thing, $url, None, ParseOptions::DEFAULT_XML)),
  ...
}
```

so then you could incrementally omit arguments.

I tend to prefer the struct approach as the compiler is extremely helpful, but with macros you could really get close to Ruby-level shorthands.

All that said, shorthands aren't necessarily a big concern to me for now. I'm mostly worried about "ease of use" for sophisticated examples, such as creating and rewriting entire XML trees, ease of use of namespaces and xpath (which are the 2 trickiest beasts) and ease of injecting into native Rust idioms (such as the Rayon technique I mentioned in #20 ). Will try to set aside some time for hammering down some docs and tests, github comments are way easier...

**Comment by triptec** [comment]
Some docs and tests sounds great. What timeframe are we talking about? I’m hoping to get some free time tonight or tomorrow and play with something and without any docs I’ll just try and copy Nokogiri::XML api and see how that feels

**Comment by dginev** [comment]
To put a face on this issue, here is the bindgen-generated file for libxml2:
https://github.com/KWARC/rust-libxml/blob/bindgen/src/bindings.rs

My main idea for a demo PR would be to replace the `c_signatures.rs` file with the `bindings.rs` one, and get to parity with master in passing tests, etc.

Example of what we have in c_signatures is:
```rust
pub fn xmlAddChild(cur: *mut c_void, new: *const c_void) -> *mut c_void;
```

which looks equally as reasonable in the bindgen-produced file:
```rust
extern "C" {
  pub fn xmlAddChild(parent: xmlNodePtr, cur: xmlNodePtr) -> xmlNodePtr;
}
```

where `xmlNodePtr` is defined as:
```rust
/// xmlNode:
///
/// A node in an XML tree.
pub type xmlNode = _xmlNode;
pub type xmlNodePtr = *mut xmlNode;
#[repr(C)]
#[derive(Debug, Copy, Clone)]
pub struct _xmlNode {
  pub _private: *mut ::std::os::raw::c_void,
  pub type_: xmlElementType,
  pub name: *const xmlChar,
  pub children: *mut _xmlNode,
  pub last: *mut _xmlNode,
  pub parent: *mut _xmlNode,
  pub next: *mut _xmlNode,
  pub prev: *mut _xmlNode,
  pub doc: *mut _xmlDoc,
  pub ns: *mut xmlNs,
  pub content: *mut xmlChar,
  pub properties: *mut _xmlAttr,
  pub nsDef: *mut xmlNs,
  pub psvi: *mut ::std::os::raw::c_void,
  pub line: ::std::os::raw::c_ushort,
  pub extra: ::std::os::raw::c_ushort,
}
```

one interesting difference is that we switch from a black box interface to libxml2, to a white one, where we can directly see into and access the `xmlNode` struct in Rust. I'm tempted to give this a spin...

**Comment by dginev** [comment]
I now opened a PR #39 , since I managed to swap in the bindgen-generated functions surprisingly quickly.

**Comment by dginev** [comment]
Thank you for bringing the bindgen approach to our attention @FranklinChen , and even bigger thanks for sharing your files and setup which helped me get things working smoothly very quickly.

The PR is now merged, and I am about to release a 0.2.1 version that includes the full bindgen signatures on crates.io, so feel free to give that a spin if you are still interested. Thanks!

**Comment by FranklinChen** [comment]
@dginev Sounds great, I may retrofit an old project to use this!
